src/docker/composeappengine.cc

```diff
@@ -193,4 +320,77 @@ void ComposeAppEngine::extractAppArchive(const App& app, const std::string& arch
   }
 }
 
+ComposeAppEngine::AppState::AppState(const App& app, const boost::filesystem::path& root, bool store_version) try
+    : version_file_ {
+  (root / MetaDir / VersionFile).string()
+}
+, state_file_{(root / MetaDir / StateFile).string()} {
+  version_ = version_file_.readStr();
```

Comment:
i'm guessing you've thought through this, but its sort of feeling like this would be easier and more atomic if we had a single state file that had both values.

Reply:
Not sure so. First of all, a version changes only once, while there are many states, so writing hash/URL on each state update is kind of useless. Also, writing both URI and state into a single file makes business logic slightly more complicated.
In terms of atomicity, it is atomic from AppState user perspective.